crates/proof-of-sql/src/sql/proof_exprs/range_check.rs

```diff
@@ -0,0 +1,154 @@
+use crate::base::{scalar::Scalar, slice_ops};
+use bumpalo::Bump;
+use bytemuck::cast_slice;
+
+fn decompose_scalar_to_words<'a, S: Scalar + 'a>(
+    scalars: &mut [S],
+    alloc: &'a Bump,
+    word_columns: &mut [&mut [u8]],
+    byte_counts: &mut [i64],
+    all_scalar_bytes: &mut Vec<&'a [u8]>,
+) {
+    for (i, scalar) in scalars.iter().enumerate() {
+        let scalar_array: [u64; 4] = (*scalar).into(); // Convert scalar to u64 array
+        let scalar_bytes_full = cast_slice::<u64, u8>(&scalar_array); // Cast u64 array to u8 slice
+        let scalar_bytes = alloc.alloc_slice_copy(&scalar_bytes_full[..31]); // Limit to 31 bytes and allocate in bumpalo
+
+        // Populate the rows of the words table with decomposition of scalar:
+        // word_columns:
+        //
+        // | Column i           | Column i+1             | Column i+2            | ... | Column_||word||     |
+        // |--------------------|------------------------|-----------------------|-----|---------------------|
+        // | Byte i of Scalar i | Byte 1+1 of Scalar i+1 | Byte 1+2 of Scalar i+2| ... | Byte n of Scalar n  |
+        for (row, &byte) in word_columns.iter_mut().zip(scalar_bytes.iter()) {
+            row[i] = byte;
+            byte_counts[byte as usize] += 1;
+        }
+
+        // Store the byte array slice for use in Phase 2
+        all_scalar_bytes.push(scalar_bytes);
+    }
+}
+
+fn get_logarithmic_derivative<'a, S: Scalar + 'a>(
+    all_scalar_bytes: &[&[u8]],
+    alpha: S,
+    inverted_word_columns: &mut [&mut [S]],
+) {
+    // Phase 2: Use the stored byte arrays and alpha
+    for (i, scalar_bytes) in all_scalar_bytes.iter().enumerate() {
```

Comment:
I was told initially we just wanted something that worked and then we would come back with rayon later. I can add it in if it wont increase the size of this PR too much. I was planning to come back with threaded enhancements later

Reply:
Right. PR review is a reasonable place for us to add in this perf improvement since it's so simple.

Reply:
Im getting non-trivial synchronization problems when trying this approach, so Im going to leave rayon for a fast-follow PR for now:

```
error[E0277]: `Cell<NonNull<bumpalo::ChunkFooter>>` cannot be shared between threads safely
   --> crates/proof-of-sql/src/sql/proof_exprs/range_check.rs:15:45
    |
15  |       scalars.par_iter().enumerate().for_each(|(i, scalar)| {
    |                                      -------- ^------------
    |                                      |        |
    |  ____________________________________|________within this `{closure@crates/proof-of-sql/src/sql/proof_exprs/range_check.rs:15:45: 15:58}`
    | |                                    |
    | |                                    required by a bound introduced by this call
16  | |         let scalar_array: [u64; 4] = (*scalar).into(); // Convert scalar to u64 array
17  | |         let scalar_bytes_full = cast_slice::<u64, u8>(&scalar_array); // Cast u64 array to u8 slice
18  | |         let scalar_bytes = alloc.alloc_slice_copy(&scalar_bytes_full[..31]); // Limit to 31 bytes and allocate in bumpalo
...   |
34  | |         }
35  | |     });
    | |_____^ `Cell<NonNull<bumpalo::ChunkFooter>>` cannot be shared between threads safely
    |
    = help: within `{closure@crates/proof-of-sql/src/sql/proof_exprs/range_check.rs:15:45: 15:58}`, the trait `Sync` is not implemented for `Cell<NonNull<bumpalo::ChunkFooter>>`, which is required by `{closure@crates/proof-of-sql/src/sql/proof_exprs/range_check.rs:15:45: 15:58}: Sync`
    = note: if you want to do aliasing and mutation between multiple threads, use `std::sync::RwLock`
note: required because it appears within the type `Bump`
   --> /home/debian/.cargo/registry/src/index.crates.io-6f17d22bba15001f/bumpalo-3.16.0/src/lib.rs:292:12
    |
292 | pub struct Bump {
    |            ^^^^
    = note: required because it appears within the type `&Bump`
note: required because it's used within this closure
   --> crates/proof-of-sql/src/sql/proof_exprs/range_check.rs:15:45
    |
15  |     scalars.par_iter().enumerate().for_each(|(i, scalar)| {
    |                                             ^^^^^^^^^^^^^
note: required by a bound in `rayon::iter::ParallelIterator::for_each`
   --> /home/debian/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rayon-1.10.0/src/iter/mod.rs:384:30
    |
382 |     fn for_each<OP>(self, op: OP)
    |        -------- required by a bound in this associated function
383 |     where
384 |         OP: Fn(Self::Item) + Sync + Send,
    |                              ^^^^ required by this bound in `ParallelIterator::for_each`
    ```